sklearn/utils/_array_api.py

```diff
@@ -299,10 +299,18 @@ def supported_float_dtypes(xp):
 
     https://data-apis.org/array-api/latest/API_specification/data_types.html
     """
+    dtypes_dict = xp.__array_namespace_info__().dtypes(
+        kind="real floating", device=device
+    )
+    valid_float_dtypes = []
+    for dtype_key in ("float64", "float32"):
+        if dtype_key in dtypes_dict:
+            valid_float_dtypes.append(dtypes_dict[dtype_key])
+
     if hasattr(xp, "float16"):
-        return (xp.float64, xp.float32, xp.float16)
-    else:
-        return (xp.float64, xp.float32)
+        valid_float_dtypes.append(xp.float16)
```

Comment:
Is this still needed? I think it can be wrong: some devices might not support `float16` even when the namespace exposes it.

Reply:
The array API specification does not support float16 which is why we have this condition. https://data-apis.org/array-api/latest/API_specification/data_types.html